File: python_implementation/models/sale.py
from models import Base
from sqlalchemy import Column, Integer, DateTime, ForeignKey, CheckConstraint
from sqlalchemy.orm import relationship
from models.movie import Movie
from models.theater import Theater


class Sale(Base):
    __tablename__ = "sales"

    date = Column(DateTime, primary_key=True)
    movie_id = Column(Integer, ForeignKey("movies.id"), primary_key=True)
    movie = relationship(Movie)
    theater_id = Column(Integer, ForeignKey("theaters.id"), primary_key=True)
    theater = relationship(Theater)
    tickets_sold = Column(Integer, nullable=False)

    __table_args__ = (
        CheckConstraint(tickets_sold >= 0, name="check_tickets_sold_non_negative"),
    )

File: javascript_implementation/models/sale.js
const { DataTypes } = require("sequelize");
const sequelize = require("./db");
const Movie = require("./movie");
const Theater = require("./theater");

const Sale = sequelize.define("sales", {
  date: {
    type: DataTypes.DATEONLY,
    allowNull: false,
  },
  movie_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: Movie,
      key: "id",
    },
  },
  theater_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: Theater,
      key: "id",
    },
  },
  tickets_sold: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 0,
    },
  },
});

// Define associations between models
Sale.belongsTo(Movie, { foreignKey: "movie_id" });
Movie.hasMany(Sale, { foreignKey: "movie_id" });

Sale.belongsTo(Theater, { foreignKey: "theater_id" });
Theater.hasMany(Sale, { foreignKey: "theater_id" });

module.exports = Sale;
